subprojects/core/src/main/java/org/gradle/caching/local/internal/DirectoryBuildCacheServiceFactory.java

```diff
@@ -49,37 +49,33 @@ public class DirectoryBuildCacheServiceFactory implements BuildCacheServiceFacto
 
     private final UnscopedCacheBuilderFactory unscopedCacheBuilderFactory;
     private final GlobalScopedCacheBuilderFactory cacheBuilderFactory;
-    private final PathToFileResolver resolver;
     private final FileAccessTimeJournal fileAccessTimeJournal;
     private final CacheConfigurationsInternal cacheConfigurations;
     private final CacheCleanupStrategyFactory cacheCleanupStrategyFactory;
+    private final ProviderFactory providerFactory;
 
     @Inject
     public DirectoryBuildCacheServiceFactory(
         UnscopedCacheBuilderFactory unscopedCacheBuilderFactory,
         GlobalScopedCacheBuilderFactory cacheBuilderFactory,
-        PathToFileResolver resolver,
         FileAccessTimeJournal fileAccessTimeJournal,
         CacheConfigurationsInternal cacheConfigurations,
-        CacheCleanupStrategyFactory cacheCleanupStrategyFactory
+        CacheCleanupStrategyFactory cacheCleanupStrategyFactory,
+        ProviderFactory providerFactory
     ) {
         this.unscopedCacheBuilderFactory = unscopedCacheBuilderFactory;
         this.cacheBuilderFactory = cacheBuilderFactory;
-        this.resolver = resolver;
         this.fileAccessTimeJournal = fileAccessTimeJournal;
         this.cacheConfigurations = cacheConfigurations;
         this.cacheCleanupStrategyFactory = cacheCleanupStrategyFactory;
+        this.providerFactory = providerFactory;
     }
 
     @Override
     public BuildCacheService createBuildCacheService(DirectoryBuildCache buildCacheConfig, Describer describer) {
-        Object cacheDirectory = buildCacheConfig.getDirectory();
-        File target;
-        if (cacheDirectory != null) {
-            target = resolver.resolve(cacheDirectory);
-        } else {
-            target = cacheBuilderFactory.baseDirForCrossVersionCache(BUILD_CACHE_KEY);
-        }
+        File target = buildCacheConfig.getDirectory().getAsFile().orElse(
```

Comment:
❓ Could that be just:
```
buildCacheConfig.getDirectory().getAsFile().getOrElse(cacheBuilderFactory.baseDirForCrossVersionCache(BUILD_CACHE_KEY))
```
or is `cacheBuilderFactory.baseDirForCrossVersionCache` expensive?

Reply:
To me no solution is ideal here, we should have:
```
buildCacheConfig.getDirectory().getAsFile().getOrElse(() -> cacheBuilderFactory.baseDirForCrossVersionCache(BUILD_CACHE_KEY))
``` 
:) 

What if we just do:
```
File target = buildCacheConfig.getDirectory().getAsFile().getOrNull()
```
and we keep old logic the same?

Reply:
> we should have

Yep, this was my initial intent. Unfortunately, there is no support for this:(

> What if we just do

Okay with me. I agree that there is no great solution for this for now.